File: Lab7/deadlock.py
n = 5
r = 3

# Allocation
alloc = [
    [0, 0, 1],  # P0
    [3, 0, 0],  # P1
    [1, 0, 1],  # P2
    [2, 3, 2],  # P3
    [0, 0, 3]   # P4
]
# Max
maxm = [
    [7, 6, 3],  # P0
    [3, 2, 2],  # P1
    [8, 0, 2],  # P2
    [2, 1, 2],  # P3
    [5, 2, 3]   # P4
]
# Available
avail = [2, 3, 2]

f = [0] * n        # finished
ans = [0] * n      # safe sequence
ind = 0

# Need
need = [[0] * r for _ in range(n)]
for i in range(n):
    for j in range(r):
        need[i][j] = maxm[i][j] - alloc[i][j]

# Banker's Algorithm
for _ in range(n):
    for i in range(n):
        if f[i] == 0:
            flag = 0
            for j in range(r):
                if need[i][j] > avail[j]:
                    flag = 1
                    break

            if flag == 0:
                ans[ind] = i
                ind += 1
                for y in range(r):
                    avail[y] += alloc[i][y]
                f[i] = 1

print("The SAFE Sequence is as follows")
for i in range(n - 1):
    print(f"P{ans[i]} ->", end=" ")
print(f"P{ans[n - 1]}")

File: LabFinal/deadlock.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    int n, r;

    cout << "Enter number of processes: ";
    cin >> n;

    cout << "Enter number of resources: ";
    cin >> r;

    vector<vector<int>> alloc(n, vector<int>(r));
    vector<vector<int>> maxm(n, vector<int>(r));
    vector<int> avail(r);

    // Allocation matrix input
    cout << "Enter Allocation Matrix:\n";
    for (int i = 0; i < n; i++) {
        cout << "P" << i << ": ";
        for (int j = 0; j < r; j++) {
            cin >> alloc[i][j];
        }
    }

    // Max matrix input
    cout << "Enter Max Matrix:\n";
    for (int i = 0; i < n; i++) {
        cout << "P" << i << ": ";
        for (int j = 0; j < r; j++) {
            cin >> maxm[i][j];
        }
    }

    // Available resources
    cout << "Enter Available Resources: ";
    for (int j = 0; j < r; j++) {
        cin >> avail[j];
    }

    // --- Banker's Algorithm starts ---
    vector<vector<int>> need(n, vector<int>(r));
    vector<int> finished(n, 0), safeSeq;

    // Need = Max - Allocation
    for (int i = 0; i < n; i++)
        for (int j = 0; j < r; j++)
            need[i][j] = maxm[i][j] - alloc[i][j];

    // Safety check
    for (int k = 0; k < n; k++) {
        for (int i = 0; i < n; i++) {
            if (!finished[i]) {
                bool canRun = true;

                for (int j = 0; j < r; j++) {
                    if (need[i][j] > avail[j]) {
                        canRun = false;
                        break;
                    }
                }

                if (canRun) {
                    safeSeq.push_back(i);

                    for (int j = 0; j < r; j++)
                        avail[j] += alloc[i][j];

                    finished[i] = 1;
                }
            }
        }
    }

    // Output
    if (safeSeq.size() == n) {
        cout << "Safe Sequence: ";
        for (int i = 0; i < n; i++) {
            cout << "P" << safeSeq[i];
            if (i != n - 1) cout << " -> ";
        }
        cout << endl;
    } else {
        cout << "System is NOT in a safe state\n";
    }

    return 0;
}